### DISCUSSION TITLE
Questions about liquidity calculation

### DISCUSSION THREAD

**Discussion by ppoliani** [Jeiwan/uniswapv3-book#60] [Milestone 1. First Swap] [answered]
First of all, I'd like to acknowledge that this is probably the best material there exist regarding Uniswap V3. The book is just amazing.

Regarding the question. I find it quite hard to grasp the section that describes the calculation of L. More specifically:

> That being said, we want to find an L that will allow the price to move to either of the points. We want enough liquidity for the price to reach either of the boundaries of a price range

I'm not sure how to interpret this to, be honest. Why do we want to do that. What is our goal here. I mean why do we start calculating L based on this fact?

The next part which again I'm struggling to understand is the two segments (Left and right).

>  Because each of the two tokens of a pool contributes to either of the segments: the left segment is made entirely of token x, the right segment is made entirely of token y. 

I think the confusion stems from the fact that I find it hard to to conceptualise points b and a. What's the relation of these two points to the the the price boundaries in your example i.e. 4545 and 5500?

> Thus, the liquidity in the segment of the curve to the left of the current price consists only of token x and is calculated only from the amount of token x provided. 

Does't the range between b and a include the liquidity that we're adding i.e. 1 ETH and 5000 USDC which means the entire area of that curve includes both tokens?

Sorry for asking these questions, but I feel like this part if super vital to understand the remaining of the book and I'm kind of stuck at this point.

**Comment by Jeiwan** [comment, marked as answer]
> I mean why do we start calculating L based on this fact?

This comes from the nature of the constant product formula, $x * y = k$: when liquidity is added, it must be spread evenly along the entire curve. This guarantees that there's liquidity at any point of the curve (at any price that's within the curve). Uniswap V2 used an infinite curve (it never crossed the axes) but Uniswap V3 uses curves that are limited by user-chosen prices. Thus, when a liquidity provider adds liquidity, it must be guaranteed that the liquidity is available at any price within the chosen price range, but not at the boundaries or outside of the price range.

> What's the relation of these two points to the the the price boundaries in your example i.e. 4545 and 5500?

Points `a` and `b` are the boundaries of a price range chosen by a liquidity provider. The curve is built between these points (different price ranges will have different curves, with different amount of liquidity). Liquidity provided between points `a` and `b` is spread evenly along the curve (proportionally to the price) so that, at any point of the curve, there are reserves of both tokens. But in points `a` and `b` only either of the two tokens is available for swapping. Thus, swapping is technically not possible at this points and the price will move into a neighbor liquidity position.

> Does't the range between b and a include the liquidity that we're adding i.e. 1 ETH and 5000 USDC which means the entire area of that curve includes both tokens?

Yes, but not in points `a` and `b`. These are the points where the curve touches either of the axes, thus the other axis is zero in these points and thus the reserve of either token is also zero. When the current price reaches either of the points, it must exit this price range and move into another price range that has liquidity above or below the touched point.

**Reply by ppoliani** [reply]
@Jeiwan Thank you so much for your detailed answer :)
